go/vt/workflow/manager.go

```diff
@@ -0,0 +1,374 @@
+package workflow
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+	gouuid "github.com/satori/go.uuid"
+	"golang.org/x/net/context"
+
+	"github.com/youtube/vitess/go/vt/topo"
+
+	workflowpb "github.com/youtube/vitess/go/vt/proto/workflow"
+)
+
+var (
+	// factories has all the factories we know about.
+	factories = make(map[string]Factory)
+)
+
+// Workflow is a running instance of a job.
+type Workflow interface {
+	// Run runs the Workflow within the provided WorkflowManager.
+	// It should return ctx.Err() if ctx.Done() is closed.  The
+	// Run method can alter the provided WorkflowInfo to save its
+	// state (and it can checkpoint that new value by saving it
+	// into the manager's topo Server).
+	Run(ctx context.Context, manager *Manager, wi *topo.WorkflowInfo) error
```

Comment:
optional: Instead of passing `Manager`, you could be more explicit and for example pass in the `NodeManager` and an interface which allows to checkpoint the state.